modules/javafx.graphics/src/main/native-glass/gtk/glass_window_ime.cpp

```diff
@@ -23,266 +23,169 @@
 }
 
-//Note: this function must return int, despite the fact it doesn't conform to XIMProc type.
-// This is required in documentation of XIM
-static int im_preedit_start(XIM im_xim, XPointer client, XPointer call) {
-    (void)im_xim;
-    (void)call;
-
-    mainEnv->CallVoidMethod((jobject) client, jViewNotifyPreeditMode, JNI_TRUE);
-    CHECK_JNI_EXCEPTION_RET(mainEnv, -1);
-    return -1; // No restrictions
+static void on_commit(GtkIMContext *im_context, gchar* str, gpointer user_data) {
+    WindowContext *ctx = (WindowContext *) user_data;
+    ctx->commitIME(str);
 }
 
-static void im_preedit_done(XIM im_xim, XPointer client, XPointer call) {
-    (void)im_xim;
-    (void)call;
-
-    mainEnv->CallVoidMethod((jobject) client, jViewNotifyPreeditMode, JNI_FALSE);
-    CHECK_JNI_EXCEPTION(mainEnv);
+// Note: JavaFX did not have surround support at this time
+static gboolean on_delete_surrounding(GtkIMContext* self, gint offset, gint n_chars, gpointer user_data) {
+    return TRUE;
 }
 
-static void im_preedit_draw(XIM im_xim, XPointer client, XPointer call) {
-    (void)im_xim;
-    (void)call;
-
-    XIMPreeditDrawCallbackStruct *data = (XIMPreeditDrawCallbackStruct*) call;
-    jstring text = NULL;
-    jbyteArray attr = NULL;
-
-    if (data->text != NULL) {
-        if (data->text->string.multi_byte) {
-            if (data->text->encoding_is_wchar) {
-                size_t csize = wcstombs(NULL, data->text->string.wide_char, 0);
-                char *ctext = new char[csize + 1];
-                wcstombs(ctext, data->text->string.wide_char, csize + 1);
-                text = mainEnv->NewStringUTF(ctext);
-                delete[] ctext;
-                CHECK_JNI_EXCEPTION(mainEnv);
-            } else {
-                text = mainEnv->NewStringUTF(data->text->string.multi_byte);
-                CHECK_JNI_EXCEPTION(mainEnv);
-            }
-        }
+static gboolean on_retrieve_surrounding(GtkIMContext* self, gpointer user_data) {
+    return TRUE;
+}
 
-        if (XIMFeedback* fb = data->text->feedback) {
-            attr = mainEnv->NewByteArray(data->text->length);
-            CHECK_JNI_EXCEPTION(mainEnv)
-            jbyte v[data->text->length];
-            for (int i = 0; i < data->text->length; i++) {
-                if (fb[i] & XIMReverse) {
-                    v[i] = com_sun_glass_ui_View_IME_ATTR_TARGET_NOTCONVERTED;
-                } else if (fb[i] & XIMHighlight) {
-                    v[i] = com_sun_glass_ui_View_IME_ATTR_TARGET_CONVERTED;
-                } else if (fb[i] & XIMUnderline) {
-                    v[i] = com_sun_glass_ui_View_IME_ATTR_CONVERTED;
-                } else {
-                    v[i] = com_sun_glass_ui_View_IME_ATTR_INPUT;
-                }
-            }
-            mainEnv->SetByteArrayRegion(attr, 0, data->text->length, v);
-            CHECK_JNI_EXCEPTION(mainEnv)
-        }
+void WindowContextBase::commitIME(gchar *str) {
+    if (im_ctx.on_preedit) {
```

Comment:
Though `LOG0`/`LOG1`, I found that the value of `im_ctx.on_preedit` is `false` under `fcitx`.
I removed this check, then tested under `fcitx` and `ibus`. All passed.

Reply:
@hellyguo Could you provide more details on which Linux distribution, input method, and fcitx version you're using?

I just tested this on Ubuntu 24.10. For Chinese I'm using Intelligent Pinyin and for Japanese I'm using Mozc. Everything worked under Ibus but when I switched to fcitx 4 both my Chinese and Japanese input methods stopped working for all apps, not just JavaFX. This makes me think that I'm not setting up fcitx correctly. I did reboot at each step as I changed the configuration.

Reply:
hi @beldenfox / @tsayao , thanks for response.

here is my env.

```console
# uname -a
Linux workdeb 6.1.0-28-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.119-1 (2024-11-22) x86_64 GNU/Linux

# cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
NAME="Debian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

# fcitx5 -v
5.0.21  
```

![Screenshot from 2024-12-10 17-50-02](https://github.com/user-attachments/assets/c3b6139c-34d1-4422-9a6c-b8d06e4988bc)

Reply:
@hellyguo I can't reproduce the problem you're seeing. I configured a Debian Bookworm system with Fcitx5 and Pinyin and tested this PR and everything worked. The Linux and Fcitx version numbers are the same as yours. I also installed fcitx5 on my existing Ubuntu 24.10 system and it worked there, also.

The behavior seems a little odd but it matches the behavior of other Linux apps (I see Roman letters inserted into the JavaFX TextField until I select Chinese text in the IM). In your screen capture the input method window was clearly in the wrong location but in my testing it was correctly positioned near the JavaFX caret.

Reply:
Finally, I got the point. Fcitx5 has a setting entry: `Show preedit in application`. If this is on, it works well. If this is off, the value `im_ctx.on_preedit` will get `false` when im(`fcitx5`) finished input-action. Then, the input will not show in the TextField.

![wechat_2024-12-12_193021_345](https://github.com/user-attachments/assets/9cab3844-8166-43fe-b69d-bda00f422e90)

Reply:
![pic_20241212193112](https://github.com/user-attachments/assets/7b9233f9-3af5-42d5-89f9-9f85248d68dd)

The option is surrounded by a red box.

Reply:
@hellyguo Thanks so much for tracking this down and providing all the details. I would like to better understand what that checkbox does but I might not be able to get to that until next week.

JavaFX handles keyboard entry using two mechanisms, KeyEvents and InputMethodEvents. This corresponds to the way Windows works. On Mac and Linux the OS only uses one channel (basically InputMethodEvents) and we have to apply a simple heuristic to map this to JavaFX's two-channel approach. This checkbox seems to be upsetting that logic.

(I work on the Mac side of this and @tsayao works on the Linux side. I just do testing on Linux because it's easy for me to create new virtual machines.)

Reply:
Anyway, current solution without checking `im_ctx.on_preedit` can work well. I backported it to `jfx22` and it worked well. Thanks again. And hope for the final solution.

Reply:
This option sets if editing happens on the application or in the IME window.

Application:
![image](https://github.com/user-attachments/assets/d44364ad-4090-4e1d-9fcf-5fa7fbd8a306)


IME Window:
![image](https://github.com/user-attachments/assets/692536bf-e8d2-467d-9332-0a6f04f1eb9a)

Reply:
I modified it to no longer rely on pre-edit calls, but I'm not sure if it's working as expected yet.

Reply:
When fcitx is configured this way we never receive pre-editing signals AND we're not filtering a key event when the commit comes in. The best way to test this is to use the mouse to click on a choice in the IM window. You'll see a commit arrive even though no key was pressed (which is expected) and no pre-edit signals were sent (which is unexpected).

On the one hand according to the GTK documentation this set of conditions should never happen. On the other hand this is Linux.

I'll look at this code in more detail but I think you're comparing the commit text with the keyval of the previous key event which may be unrelated. The simplest solution is to add a flag that's only set while a key event is being filtered. If a commit message arrives during pre-edit OR when no key event is being filtered it should be processed as an InputMethod event. I did a quick test of this logic and it seems to work but I did very little testing.

Reply:
I reverted and did the way @beldenfox explained above.